functions/src/services/GoogleSheet.ts

```diff
@@ -0,0 +1,292 @@
+const gapi = require('googleapis').google;
+
+export enum GoogleScopes {
+  SpreadSheets = 'https://www.googleapis.com/auth/spreadsheets',
+}
+
+type ISheetId = string;
+
+// Create other SpreadSheet Interfaces and add to IProjectSpreadSheetNames
+// enum IOtherReportSheet {  }
+
+export type IProjectSpreadSheetNames = string;
+export type ISheetRowTypes = any;
+
+enum IMajorDimensions {
+  Rows = 'ROWS',
+  Columns = 'COLUMNS',
+}
+
+enum IValueInputOption {
+  USER_ENTERED = 'USER_ENTERED',
+  RAW = 'RAW',
+}
+
+interface IValueRange {
+  range: string;
+  majorDimension: IMajorDimensions;
+  values: IRowValues[];
+}
+
+interface IGetSheetResponse {
+  headers: any[];
+  rows: any[];
+}
+
+abstract class SpreadSheet {
+  protected abstract connect(): any;
+  public abstract getSheet(
+    spreadsheetId: ISheetId,
+    sheetName: IProjectSpreadSheetNames,
+    limit?: number
+  ): Promise<IGetSheetResponse>;
+  public abstract appendRow(
+    spreadsheetId: ISheetId,
+    sheetName: IProjectSpreadSheetNames,
+    appendValues: any
+  ): Promise<any>;
+  public abstract getColumn(
+    spreadsheetId: ISheetId,
+    sheetName: IProjectSpreadSheetNames,
+    columnName: string
+  ): Promise<any>;
+  public abstract findRowWithColumnValue(
+    column: string,
+    rowValueToSearch: string | any,
+    tableToSearch: any
+  ): Promise<any>;
+  public abstract updateAllotmentRow(
+    sheetId: string,
+    sheetName: IProjectSpreadSheetNames,
+    rowNumber: number,
+    updateValues: any
+  ): Promise<any>;
+}
+
+type IRowValues = string[] | number[];
+
+export default class GoogleSheets extends SpreadSheet {
+  protected connection: any;
+
+  protected async connect() {
```

Comment:
You are calling this method from each other method. Shouldn't we add some protection from connecting again and again?

Reply:
Return condition added.